in-progress/6973-refactor-base-contract-interaction.md

```diff
@@ -0,0 +1,1092 @@
+  // will go away soon.
+  await this.getDeploymentFunctionCalls(options);
+  request.fee = await this.getFeeOptionsFromEstimatedGas(request);
+}
+```
+
+We also want to unify the mechanism for creating, simulating, proving, and sending transactions, regardless of whether they are for deploying a contract, calling a function, or deploying an account.
+
+This would clear up concerns like:
+
+> // REFACTOR: Having a `request` method with different semantics than the ones in the other
+  // derived ContractInteractions is confusing. We should unify the flow of all ContractInteractions.
+
+
+## Interface
+
+The key operations we want to perform are:
+- Simulating/Reading
+- Proving
+- Submitting
+
+These operations should be accessible from a `Wallet`, which will generally be `Schnorr` or `Signerless`. 
+
+This is analogous to `viem` having a `walletClient` and a `publicClient`.
+
+The main idea in this design is to have the BaseWallet define all the functionality to perform the key operations; the specific entrypoint in use will be able to adapt the `TxExecutionRequest`s produced to its needs.
+
+To accomplish this, we have the BaseWallet construct a `TxExecutionRequestBuilder` that can be passed to various `TxExecutionRequestAdapters`, such as one for the account entrypoint, one for the fee payment method, etc.
+
+Further, we expand `TxExecutionRequest` to allow passing things that we were previously setting directly on the PXE, so simulations can be idempotent.
+
+The interfaces are as follows:
+
+```ts
+
+// new
+export interface ArtifactAndInstance {
+  artifact: ContractArtifact;
+  instance: ContractInstanceWithAddress;
+}
+
+// modified
+export class TxExecutionRequest {
+  constructor(
+    // All these are the same:
+    public origin: AztecAddress,
+    public functionSelector: FunctionSelector,
+    public firstCallArgsHash: Fr,
+    public txContext: TxContext,
+    public argsOfCalls: PackedValues[],
+    public authWitnesses: AuthWitness[],
+    // Add:
+    /**
+     * Transient capsules needed for this execution.
+     */
+    public capsules: Fr[][],
+    /**
+     * Transient contracts needed for this execution.
+     */
+    public transientContracts: ArtifactAndInstance[],
+  ) {}
+  // ...
+}
+
+// new
+export interface InstanceDeploymentParams {
+  constructorName: string;
+  constructorArgs: any;
+  salt: Fr;
+  publicKeysHash: Fr;
+  deployer: AztecAddress;
+}
+
+// new
+export interface DeploymentOptions {
+  registerClass?: boolean;
+  publicDeploy?: boolean;
+}
+
+export interface UserFunctionCall {
+  contractInstance: ContractInstanceWithAddress;
+  functionName: string;
+  args: any;
+  deploymentOptions?: DeploymentOptions;
+  contractArtifact?: ContractArtifact;
+  functionAbi?: FunctionAbi;
+}
+
+// new
+export interface UserRequest {
+  calls: UserFunctionCall[];
+  gasSettings?: GasSettings;
+  paymentMethod?: FeePaymentMethod;
+  from?: AztecAddress;
+  simulatePublicFunctions?: boolean;
```

Comment:
Quick thought. Is it actually useful to have this option? In my mind, you either simulate the full thing and not at all. I get we need some simulation for proving still, but I guess I would remove the option entirely and then have them rely on using `simulate` to simulate instead of the proving as well 🤷. So if you wanna know you do simulate + prove, but if you just try to blast as quickly as possible you will just do prove.

Reply:
Yeah I agree. It is in there because it was there already. If no one has strong opinions otherwise, I'd say we nuke it.